File: CivitasRuby/lib/mazo_sorpresas.rb
#encoding:utf-8
require_relative "sorpresa"
require_relative "diario"

module Civitas

  class MazoSorpresas
       
    attr_reader :sorpresas , :usadas , :ultima_sorpresa
    
    def initialize(dbug=false)
     @barajada = false
     @usadas = 0
     @sorpresas = Array.new
     @cartas_especiales = Array.new
     @debug = dbug
     @ultima_sorpresa = nil
     
     if(@debug)
       Diario.instance.ocurre_evento("Se ha creado un mazo en modo depuración")
     end
    end
    
    
    def to_string
      i = 0
      resultado = "Sorpresas:\n"
      for sorpresa in @sorpresas
        resultado = resultado + sorpresa.to_string
      end
      resultado
    end
   

    
    def init
      @sorpresas = Array.new
      @ultima_sorpresa = nil
      @cartasEspeciales = Array.new
      @usadas = 0
      @barajada = false
    end
    
    def alMazo(sorpresa)
      if(!@barajada)
        @sorpresas<<sorpresa
      end
    end
    
    def siguiente
        
      if(!@barajada or @sorpresas.length == @usadas)
          if(!@debug)
            @sorpresas.shuffle!
          end
          
          @usadas = 0
          @barajada = true
        end
        
        @usadas += 1
        
        primera = @sorpresas.shift
        @sorpresas.push(primera)  
        @ultima_sorpresa = primera
        
        
        @ultima_sorpresa
    end
    
    def inhabilitarCartaEspecial(sorpresa)
      
      for i in (0..@sorpresas.length - 1)
        if @sorpresas[i] == sorpresa
          aux = @sorpresas[i]
          @sorpresas.delete_at(i)
          @cartas_especiales << aux
          Diario.instance.ocurre_evento("Se ha movido una carta de sorpresa a Especial")
        end
      end
      
    end
    
    def habilitarCartaEspecial(sorpresa)
    
      for i in (0..@cartas_especiales.length-1)
        if @catas_especiales[i] == sorpresa
          aux = @cartas_especiales[i]
          @cartas_especiales.delete_at(i)
          @sorpresas<<aux
          Diario.instance.ocurre_evento("Se ha movido una carta de Especiales a Sorpresas")
        end
      end
    end
    
    def tam_sorpresas
      @sorpreas.length
    end

  end
  
end

File: CivitasJava/src/civitas/MazoSorpresas.java
package civitas;
import java.util.ArrayList;
import java.util.Collections; //Para barajar el mazo

public class MazoSorpresas{

     private Sorpresa ultimaSorpresa; 
     private boolean barajada;
     private int usadas;
     private boolean debug;
     
     //asociaciones
     private ArrayList<Sorpresa> sorpresas; //0..*
     private ArrayList<Sorpresa> cartasEspeciales; //Almacenará la carta sorpresa SALIRCARCEL 1..*
    
     
    
     
    //paquete
    MazoSorpresas(){
         debug = false;
         this.init();     
    }
     
    MazoSorpresas(boolean debug){
         debug = debug;
         
         if(debug == true)
             Diario.getInstance().ocurreEvento("Se ha creado un mazo sorpresas en modo depuración");
         
         this.init();   
    }
     
    void alMazo(Sorpresa s){
         if(!barajada)
             sorpresas.add(s);
    }
     
    Sorpresa getUltimaSorpresa(){
         return sorpresas.get(sorpresas.size() - 1);
    }
     
    void habilitarCartaEspecial(Sorpresa especial){
        boolean resultado = cartasEspeciales.remove(especial);
        if(resultado){
            sorpresas.add(especial);
            Diario.getInstance().ocurreEvento("Se ha habilitado una carta especial");
        }
              
    }
     
    void inhabilitarCartaEspecial(Sorpresa especial){
        boolean resultado = sorpresas.remove(especial);
        if(resultado){
            cartasEspeciales.add(especial);
            Diario.getInstance().ocurreEvento("Se ha inhabilitado una carta especial");
        }     
    }
     
    Sorpresa siguiente(){
         
        if(!barajada  || usadas == sorpresas.size()){      
             if(!debug){   
                 Collections.shuffle(sorpresas);
                 barajada = true;
                 usadas = 0;  
             }
             
             usadas ++;
             //Quito la primera carta y la guardo en el atributo ultimaSorpresa
             //el método de instancia remove devuelve el elemento eliminado
             ultimaSorpresa = sorpresas.get(0);
             sorpresas.remove(0);
             //Pongo la sorpresa eliminada de la primera posicion en la última
             sorpresas.add(ultimaSorpresa);
         }
         
         return ultimaSorpresa;
    }
         
    
    void muestraSorpresas(){
        for(Sorpresa sorpresa:sorpresas)
            System.out.println(sorpresa.toString() +  " ");
    }
    
    //private
    private void init(){
         //Inicializamos las listas a un contenedor vacío , con capacidad inicial de 10
         sorpresas = new ArrayList(); 
         cartasEspeciales = new ArrayList();
         
         barajada = false;
         usadas = 0;
    }

}